typescript/src/index.ts

```diff
@@ -27,4 +31,13 @@ const SpvMaintainer = {
   submitRedemptionProof,
 }
 
-export { TBTC, SpvMaintainer, ElectrumClient, EthereumBridge }
+export {
+  Bitcoin,
+  TBTC,
+  SpvMaintainer,
+  OptimisticMinting,
+  ElectrumClient,
+  Ethereum,
+  EthereumBridge,
+  EthereumTBTCVault,
+}
```

Comment:
Although it seems convenient now, I'm afraid that exposing the whole modules (for example, `Bitcoin` and `Ethereum` here) without any control will make the API of the library unusable in the long term and will cause a lot of confusion for users. We should rather aim for a concise and well-defined interface oriented around specific use cases. 

So far the approach was:
- `TBTC` exposes everything that is needed to interact with the tbtc contracts as an user (dApp)
- `SpvMaintainer` exposes everything needed by SPV maintainers
- `ElectrumClient` is the reference Electrum-based implementation of the `bitcoin.Client`
- `EtheremBridge` is the reference Ethereum implementation of the `chain.Bridge

If we start to expose the whole `ethereum` and `bitcoin` modules, one day we will also expose the entire `deposit` module, `redemption` module, and so on. This will produce a lot of mess as we will expose the same methods by multiple namespaces. For example, we will have `TBTC. suggestDepositWallet` and `Deposit.suggestDepositWallet` - library users won't know which one should they choose.

On the other hand, it is clear that the library grows and we need to expose more than we expose now, especially some types in order to avoid importing them from `dist` on the library client side. Assuming we want to keep the API oriented around use cases and expose types in a clear way maybe we should do the following:

```
// We avoid the 'import * as X from "module.ts"' pattern

// functions for TBTC users
const TBTC = {
  calculateDepositAddress,
  suggestDepositWallet,
  revealDeposit,
  getRevealedDeposit,
  requestRedemption,
  getRedemptionRequest,
}

// functions for SPV maintainers
const SpvMaintainer = {
  submitDepositSweepProof,
  submitRedemptionProof,
}

// functions for optimistic minting maintainers
const MintingMaintainer = {
 (...)
}

// only types, no functions! We can also name it as 'Primitives' or so
const Types = {
    Bitcoin: {
        Client,
        Transaction,
        // other relevant Bitcoin primitives
    }

    Chain: {
        Bridge,
        Identifier
        // other Chain primitives that can be useful for writing implementations
    }

    Electrum: {
        Credentials,
        Client
        // other relevant Electrum types
     }

     Ethereum: {
         Address,
         Bridge,
         TBTCVault,
         // other relevant Ethereum types
     }
}

export { 
    TBTC, 
    SpvMaintainer, 
    MintingMaintainer,
    Types
}
```
What do you think @nkuba? Also curious about @pdyraga thoughts as he originally proposed the approach we use currently.

Reply:
Unfortunatelly, I found the current approach confusing from both developer and user perspectives.

By exporting the whole modules we export only the parts that are marked explicitly with `export`, they are also accessible by direct importing from `dist` files.

I've seen dapp was using direct import from `dist`, as the index was not exporting all necessary things. Also, I've noticed code duplicating in https://github.com/threshold-network/token-dashboard/tree/main/src/threshold-ts or new functions that I would expect to be implemented in this library.

It's hard to decide in a library what an actual user would need, as we cannot predict all the use cases. We should give some flexibility to a user.

I expected this library to work as its' counterpart from V1 (`tbtc.js`) - a proxy to communicate with contracts, wrapping only complex logic or helpers like retries.

I don't want to get into a broad discussion about refactoring in this PR. I'll try to update this PR to match the approach that was previously established and leave the changes for the future.

Reply:
I would also like to hear @r-czajkowski and @michalsmiarowski expectations from the dapp development perspective.

Reply:
> I don't want to get into a broad discussion about refactoring in this PR. I'll try to update this PR to match the approach that was previously established and leave the changes for the future.

Yep definitely. However, we need to discuss it in the near future. There is a strong conflict of interest between multiple use cases and we need to agree on a single approach that will be optimal for everyone. As you said, the best we can do for now is stick to the current approach as much as we can without causing a revolution and return here later. Can be leave an explicit TODO about that in the `index.ts`?

Also, leaving my thoughts regarding other parts for future discussion:

> By exporting the whole modules we export only the parts that are marked explicitly with export, they are also accessible by direct importing from dist files.

Yeah, that's true. However, the fact that something is exported outside of its module does not imply that it should be also exported from the `index.ts` by default. The `index.ts`, just as its name implies, is the place that should define a concise entry point for library users. Exposing everything through `index.ts` by default does not serve that purpose and, as I said above, will likely produce a mess that will cause even more confusion as result.

> I've seen dapp was using direct import from dist, as the index was not exporting all necessary things. Also, I've noticed code duplicating in https://github.com/threshold-network/token-dashboard/tree/main/src/threshold-ts or new functions that I would expect to be implemented in this library.

The `dist` directory is basically a container that holds all public artifacts of the library. We cannot prevent importing from it in any way. This is why I think we should create a proper `index.ts` that will emphasize the most important things. Developers can import from `dist` what they want anyway. In that context, what is the point of exporting the entire `dist` through `index.ts`? In that case, the `index.ts` is just a path shortener.

> It's hard to decide in a library what an actual user would need, as we cannot predict all the use cases. We should give some flexibility to a user.

Totally agreed. But, isn't that flexibility provided by `dist`?

Reply:
aeea14f022e3e27570434a49c2d07a98d155c60c

Reply:
Leaving that discussion open for future reference.

Reply:
a0d79b289825c321cb34108518c51a9b3faa16b5